src/lib/core/CHIPError.h

```diff
@@ -24,52 +24,158 @@
+
+    /// Test whether if @a value can be losslessly encapsulated in a `CHIP_ERROR`.
+    static constexpr bool CanEncapsulate(BaseType value) { return FitsInField(kValueLength, value); }
+
+    /// Construct a `CHIP_ERROR` encapsulating @a value inside the @a range.
+    static BaseType Encapsulate(Range range, BaseType value)
+    {
+        return MakeInteger(range, (value & MakeMask(kValueStart, kValueLength)));
+    }
+
+    /// Test whether @a error is an SDK error belonging to @a part.
+    static constexpr bool IsPart(SdkPart part, BaseType error)
+    {
+        return (error & (MakeMask(kRangeStart, kRangeLength) | MakeMask(kSdkPartStart, kSdkPartLength))) ==
+            (MakeField(kRangeStart, static_cast<BaseType>(Range::kSDK)) | MakeField(kSdkPartStart, static_cast<BaseType>(part)));
+    }
+
+private:
+    /*
+     * The representation of a CHIP_ERROR is structured so that SDK error code constants are small, in order to improve code
+     * density on embedded builds. Arm 32, Xtensa, and RISC-V can all handle 11-bit values in a move-immediate instruction.
+     * Further, `SdkPart::kCore` is 0 so that the most common errors fit in 8 bits for additional density on some processors.
+     *
+     *  31    28      24      20      16      12       8       4       0    Bit
+     *  |       |       |       |       |       |       |       |       |
+     *  |     range     |                     value                     |
+     *  |    kSdk==0    |       0               |0| part|    code       |   SDK error
+     *  |    01 - FF    |          encapsulated error code              |   Encapsulated error
+     */
+    static constexpr int kRangeStart  = 24;
+    static constexpr int kRangeLength = 8;
+    static constexpr int kValueStart  = 0;
+    static constexpr int kValueLength = 24;
+
+    static constexpr int kSdkPartStart  = 8;
+    static constexpr int kSdkPartLength = 3;
+    static constexpr int kSdkCodeStart  = 0;
+    static constexpr int kSdkCodeLength = 8;
+
+    static constexpr BaseType GetField(unsigned int start, unsigned int length, BaseType value)
+    {
+        return (value >> start) & ((1u << length) - 1);
+    }
+    static constexpr BaseType MakeMask(unsigned int start, unsigned int length) { return ((1u << length) - 1) << start; }
+    static constexpr BaseType MakeField(unsigned int start, BaseType value) { return value << start; }
+    static constexpr bool FitsInField(unsigned int length, BaseType value) { return value < (1u << length); }
+
+    static constexpr BaseType MakeInteger(Range range, BaseType value)
+    {
+        return MakeField(kRangeStart, static_cast<std::underlying_type<Range>::type>(range)) | MakeField(kValueStart, value);
+    }
+    static constexpr BaseType MakeInteger(SdkPart part, BaseType code)
+    {
+        return MakeInteger(Range::kSDK,
+                           MakeField(kSdkPartStart, static_cast<std::underlying_type<SdkPart>::type>(part)) |
```

Comment:
```suggestion
                           MakeField(kSdkPartStart, to_underlying(part)) |
```
after including `lib/support/TypeTraits.h`